File: solved/322.coin_change.rb
# @param {Integer[]} coins
# @param {Integer} amount
# @return {Integer}

def dp(coins, amount)
    return -1 if amount < 0
    return @mem[amount] if @mem.has_key?(amount)

    min = 50000
    coins.each do |coin|
        current = 1 + dp(coins, amount-coin)
        min = current if current > 0 && current < min
    end

    @mem[amount] = min
end

def coin_change(coins, amount)
    @mem = {0 => 0}
    ans = dp(coins, amount)
    ans == 50000 ? -1 : ans
end

File: solved/322.coin_change.go
package main

import (
	"fmt"
)

func coinChange(coins []int, amount int) int {
	dp := make([]int, amount + 1)
	l_coins := len(coins)
	dp[0] = 0
	for i:=1; i<=amount; i++ {
		min := -1
		for j:=0; j<l_coins; j++ {
			pos := i - coins[j]
			if pos >= 0 && dp[pos] >= 0 {
				if min < 0 {
					min = dp[pos] + 1
				} else {
					if dp[pos] + 1 < min {
						min = dp[pos] + 1
					}
				}
			}
		}
		dp[i] = min
	}
	return dp[amount]
}

func main() {
	var amount int
	fmt.Scan(&amount)

	var n_coins int
	fmt.Scan(&n_coins)

	coins := make([]int, n_coins)
	for i:=0; i<n_coins; i++ {
		fmt.Scan(&coins[i])
	}
	fmt.Println(coinChange(coins, amount))
}
